packages/eslint-plugin/src/util/astUtils.ts

```diff
@@ -4,27 +4,29 @@ import {
   TSESTree,
 } from '@typescript-eslint/experimental-utils';
 
+type NegateGuard<T, U> = T extends U ? T : U;
```

Comment:
i don't like this name, and this is not exclude or extract but it works in similar way